src/qibolab/instruments/rfsoc.py

```diff
@@ -0,0 +1,488 @@
+
+        * self.channels is a list that contain channel number and nyquist zone of initialization
+
+
+        Templates:
+        self.pulse_sequence = {
+            serial: {
+                "channel":
+                "type":
+                "freq":
+                "length":
+                "phase":
+                "time":
+                "gain"
+                "waveform":     # readout as a default value
+
+                "shape":    # these only if drive
+                "sigma":
+                "name":
+
+                "delta":    # these only if drag
+                "alpha":
+
+                "adc_trig_offset":      # these only if readout
+                "wait": False
+                "syncdelay": 100
+            }
+        }
+        self.readouts = {
+            0: {
+                "adc_ch":
+                "gen_ch":
+                "length":
+                "freq":
+            }
+        }
+        self.channels = [(channel, generation), ...]
+
+        """
+
+        for pulse in sequence:
+            pulse_dic = {}
+
+            pulse_dic["time"] = pulse.start
+            length = pulse.duration * self.mu_s
+            pulse_dic["length"] = self.soc.us2cycles(length)  # uses tproc clock now
+
+            if pulse.amplitude > 1:
+                raise Exception("Amplitude is relative and must be <= 1")
+            pulse_dic["gain"] = int(pulse.amplitude * self.max_gain)
+
+            if pulse.type == PulseType.DRIVE:
+                pulse_dic["type"] = "qd"
+                gen_ch, adc_ch = self.from_qubit_to_ch(pulse.qubit, "qd")
+                pulse_dic["ch"] = gen_ch
+
+                pulse_dic["freq"] = self.soc.freq2reg(pulse.frequency * self.MHz, gen_ch=gen_ch)
+
+                pulse_dic["phase"] = self.deg2reg(pulse.relative_phase, gen_ch=gen_ch)
+
+                type_p_shape = type(pulse.shape)
+                if type_p_shape is not Gaussian and type_p_shape is not Drag:
+                    raise NotImplementedError("Drive pulses can only be Gaussian or Drag")
+
+                pulse_dic["style"] = "arb"
+                sigma = length / pulse.shape.rel_sigma
+                pulse_dic["sigma"] = self.soc.us2cycles(sigma)
+
+                if type_p_shape is Gaussian:
+                    pulse_dic["waveform"] = "Gaussian"  # TODO redundancy
+                    pulse_dic["shape"] = "Gaussian"
+                    pulse_dic["name"] = "Gaussian"
+
+                if type_p_shape == Drag:
+                    pulse_dic["waveform"] = "Drag"  # TODO redundancy
+                    pulse_dic["shape"] = "Drag"
+                    pulse_dic["name"] = "Drag"
+
+                    pulse_dic["delta"] = pulse_dic["sigma"]  # TODO redundancy
+                    pulse_dic["alpha"] = pulse.beta
+
+            elif pulse.type == PulseType.READOUT:
+                type_p_shape = type(pulse.shape)
+                if type_p_shape is not Rectangular:
```

Comment:
```suggestion
                if not isinstance(pulse.shape, Rectangular):
```

Reply:
Thanks for the suggestion, I'm following it now